From 106b3b9dbb13c52aaaa628860a2aa5cadc87fab0 Mon Sep 17 00:00:00 2001
From: Patrick Delaunay <patrick.delaunay@foss.st.com>
Date: Mon, 6 Jun 2022 18:24:48 +0200
Subject: [PATCH 18/39] board: st: stm32mp2: add g_dnl_bind_fixup

Add a function to support two separate USB-PID for DFU (PID=0xdf11)
and for UsbMassStorage (PID=0x57) for STMicroelectronics boards,
as it is done for stm32mp1.

This patch is a preliminary step to activate DFU_OVER_USB and
UMS (CONFIG_CMD_USB_MASS_STORAGE) with the USB stack configuration
 CONFIG_USB_GADGET_VENDOR_NUM=0x0483
 CONFIG_USB_GADGET_PRODUCT_NUM=0x5720
 CONFIG_USB_GADGET_DOWNLOAD=y

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/u-boot.git 7d6af1f7696050daebc6f1a0ce30e90f693a5c09 board: st: stm32mp2: add g_dnl_bind_fixup]
Signed-off-by: Patrick Delaunay <patrick.delaunay@foss.st.com>
Change-Id: Ie390b8863c5b0d06d36612943d7bdfb083e4dc8e
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/u-boot/+/258344
Tested-by: Patrice CHOTARD <patrice.chotard@foss.st.com>
Reviewed-by: Patrice CHOTARD <patrice.chotard@foss.st.com>
---
 board/st/stm32mp2/stm32mp2.c | 15 +++++++++++++++
 1 file changed, 15 insertions(+)

diff --git a/board/st/stm32mp2/stm32mp2.c b/board/st/stm32mp2/stm32mp2.c
index b89c6e113f3..6cfa8a54827 100644
--- a/board/st/stm32mp2/stm32mp2.c
+++ b/board/st/stm32mp2/stm32mp2.c
@@ -10,6 +10,7 @@
 #include <env_internal.h>
 #include <fdt_support.h>
 #include <led.h>
+#include <g_dnl.h>
 #include <log.h>
 #include <misc.h>
 #include <mmc.h>
@@ -80,6 +81,20 @@ static void check_user_button(void)
 	clrsetbits_le32(TAMP_BOOT_CONTEXT, TAMP_BOOT_FORCED_MASK, BOOT_STM32PROG);
 }
 
+#ifdef CONFIG_USB_GADGET_DOWNLOAD
+#define STM32MP1_G_DNL_DFU_PRODUCT_NUM 0xdf11
+int g_dnl_bind_fixup(struct usb_device_descriptor *dev, const char *name)
+{
+	if (IS_ENABLED(CONFIG_DFU_OVER_USB) &&
+	    !strcmp(name, "usb_dnl_dfu"))
+		put_unaligned(STM32MP1_G_DNL_DFU_PRODUCT_NUM, &dev->idProduct);
+	else
+		put_unaligned(CONFIG_USB_GADGET_PRODUCT_NUM, &dev->idProduct);
+
+	return 0;
+}
+#endif /* CONFIG_USB_GADGET_DOWNLOAD */
+
 /* board dependent setup after realloc */
 int board_init(void)
 {
-- 
2.53.0

